shared-bindings/_protomatter/Protomatter.c

```diff
@@ -0,0 +1,293 @@
+    .proxy = {(mp_obj_t)&protomatter_protomatter_get_paused_obj,
+              (mp_obj_t)&protomatter_protomatter_set_paused_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+//|   .. method:: write(buf)
+//|
+//|     Transmits the color data in the buffer to the pixels so that they are shown.
+//|
+//|     The data in the buffer must be in "RGB565" format.  This means
+//|     that it is organized as a series of 16-bit numbers where the highest 5
+//|     bits are interpreted as red, the next 6 as green, and the final 5 as
+//|     blue.  The object can be any buffer, but `array.array` and `ulab.array`
+//|     objects are most often useful.
+//|
+STATIC mp_obj_t protomatter_protomatter_swapbuffers(mp_obj_t self_in) {
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    check_for_deinit(self);
+
+    _PM_convert_565(&self->core, self->bufinfo.buf, self->width);
+    _PM_swapbuffer_maybe(&self->core);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(protomatter_protomatter_swapbuffers_obj, protomatter_protomatter_swapbuffers);
+
+STATIC const mp_rom_map_elem_t protomatter_protomatter_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_deinit), MP_ROM_PTR(&protomatter_protomatter_deinit_obj) },
+    { MP_ROM_QSTR(MP_QSTR_paused), MP_ROM_PTR(&protomatter_protomatter_paused_obj) },
+    { MP_ROM_QSTR(MP_QSTR_swapbuffers), MP_ROM_PTR(&protomatter_protomatter_swapbuffers_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(protomatter_protomatter_locals_dict, protomatter_protomatter_locals_dict_table);
+
+STATIC void protomatter_protomatter_get_bufinfo(mp_obj_t self_in, mp_buffer_info_t *bufinfo) {
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    check_for_deinit(self);
+    
+    *bufinfo = self->bufinfo;
+}
+
+STATIC void protomatter_protomatter_swapbuffers_void(mp_obj_t self_in) {
+    protomatter_protomatter_swapbuffers(self_in);
+}
+
+STATIC void protomatter_protomatter_deinit_void(mp_obj_t self_in) {
+    protomatter_protomatter_deinit(self_in);
+}
+
+STATIC void protomatter_protomatter_set_brightness(mp_obj_t self_in, mp_float_t value) {
+    protomatter_protomatter_set_paused(self_in, mp_obj_new_bool(value <= 0));
+}
+
+STATIC const framebuffer_p_t protomatter_protomatter_proto = {
+    MP_PROTO_IMPLEMENT(MP_QSTR_protocol_framebuffer)
+    .get_bufinfo = protomatter_protomatter_get_bufinfo,
+    .set_brightness = protomatter_protomatter_set_brightness,
+    .swapbuffers = protomatter_protomatter_swapbuffers_void,
+    .deinit = protomatter_protomatter_deinit_void,
+};
+
+STATIC mp_int_t protomatter_protomatter_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, mp_uint_t flags) {
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    // a readonly framebuffer would be unusual but not impossible
+    if((flags & MP_BUFFER_WRITE) && !(self->bufinfo.typecode & MP_OBJ_ARRAY_TYPECODE_FLAG_RW)) {
+        return 1;
```

Comment:
Is there a macro for these buffer protocol return codes?

Reply:
This was modeled after STATIC mp_int_t array_get_buffer's conditional though confusingly I swapped the order of the two conditions.  I don't think there's a macro to encapsulate it.

Reply:
Ok, this is fine.